File: c/src/voice_xsampa.c
/*
 * voice_xsampa.c — X-SAMPA → IPA, and SentencePiece unigram encoding.
 *
 * C port of src/CircleAI.Voice/XsampaToIpa.cs and SentencePieceUnigram.cs.
 * See voice_xsampa.h for the contract, and fixtures/voice_*.json for the
 * answers this must reproduce exactly.
 */

#include "circle_ai/voice_xsampa.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ------------------------------------------------------------------------ */
/* UTF-8                                                                     */
/* ------------------------------------------------------------------------ */

/*
 * Length in bytes of the UTF-8 sequence starting at `s`.
 *
 * A continuation byte or an invalid lead returns 1 so the caller always makes
 * progress — a zero here would spin forever on malformed input, and this code
 * runs on text that arrives from a model.
 */
static size_t utf8_seq_len(const char *s)
{
    unsigned char c = (unsigned char)*s;
    if (c < 0x80) return 1;
    if ((c & 0xE0) == 0xC0) return 2;
    if ((c & 0xF0) == 0xE0) return 3;
    if ((c & 0xF8) == 0xF0) return 4;
    return 1;
}

/* Count code points in a NUL-terminated UTF-8 string. */
static size_t utf8_count(const char *s)
{
    size_t n = 0;
    while (*s) { s += utf8_seq_len(s); n++; }
    return n;
}

static char *dup_range(const char *start, size_t len)
{
    char *out = (char *)malloc(len + 1);
    if (!out) return NULL;
    memcpy(out, start, len);
    out[len] = '\0';
    return out;
}

/* ------------------------------------------------------------------------ */
/* The phone table                                                           */
/* ------------------------------------------------------------------------ */

/*
 * Every phone in the NCHLT Afrikaans dictionary, mapped to IPA.
 *
 * Derived from the corpus, not from memory: exactly the distinct phones in
 * nchlt_afr.dict, with every IPA character checked against the target voice's
 * own token table.
 *
 * Note "g" -> U+0261 LATIN SMALL LETTER SCRIPT G (UTF-8 C9 A1), NOT ASCII 'g'.
 * The voice's vocabulary carries ɡ; a plain 'g' would miss and be dropped.
 *
 * "h\\" is C for the X-SAMPA token h\, which is ɦ — the voiced glottal
 * fricative Afrikaans uses in "hond". This voice has no ɦ, only h, so voicing
 * is lost. It is the ONLY approximation in the table.
 */
static const struct { const char *xsampa; const char *ipa; } PHONES[] = {
    /* Vowels */
    {"a", "a"}, {"A:", "ɑː"}, {"A:r", "ɑːr"},
    {"E", "ɛ"}, {"O", "ɔ"}, {"@", "ə"},
    {"i", "i"}, {"u", "u"}, {"y", "y"},
    {"9", "œ"}, {"2:", "øː"}, {"{", "æ"},

    /* Diphthongs — NCHLT gives one token, the voice wants both elements. */
    {"9y", "œy"}, {"@i", "əi"}, {"@u", "əu"},
    {"i@", "iə"}, {"u@", "uə"},

    /* Consonants */
    {"b", "b"}, {"d", "d"}, {"f", "f"},
    {"g", "ɡ"},
    {"j", "j"}, {"k", "k"}, {"l", "l"},
    {"m", "m"}, {"n", "n"}, {"N", "ŋ"},
    {"p", "p"}, {"r", "r"}, {"s", "s"},
    {"S", "ʃ"}, {"t", "t"}, {"v", "v"},
    {"w", "w"}, {"x", "x"}, {"z", "z"},
    {"Z", "ʒ"},
    {"h\\", "h"},
};

static const size_t PHONE_COUNT = sizeof(PHONES) / sizeof(PHONES[0]);

static const char *lookup_phone(const char *xsampa)
{
    for (size_t i = 0; i < PHONE_COUNT; i++)
        if (strcmp(PHONES[i].xsampa, xsampa) == 0) return PHONES[i].ipa;
    return NULL;
}

static int is_blank(const char *s)
{
    for (; *s; s++)
        if (*s != ' ' && *s != '\t' && *s != '\n' && *s != '\r') return 0;
    return 1;
}

size_t circle_voice_xsampa_known_phone_count(void) { return PHONE_COUNT; }

const char *circle_voice_xsampa_known_phone(size_t index)
{
    return index < PHONE_COUNT ? PHONES[index].xsampa : NULL;
}

circle_voice_conversion circle_voice_xsampa_to_ipa(const char *const *xsampa, size_t count)
{
    circle_voice_conversion conv;
    memset(&conv, 0, sizeof(conv));
    if (count == 0) return conv;

    /* Worst case: every phone maps to a multi-code-point value. Four is beyond
     * anything in the table (the longest is A:r at three), so one allocation
     * suffices and no growth logic can get it wrong. */
    conv.ipa = (char **)calloc(count * 4 + 1, sizeof(char *));
    conv.unmapped = (char **)calloc(count + 1, sizeof(char *));
    if (!conv.ipa || !conv.unmapped) {
        circle_voice_conversion_free(&conv);
        return conv;
    }

    for (size_t i = 0; i < count; i++) {
        const char *phone = xsampa[i];
        if (!phone || is_blank(phone)) continue;

        const char *mapped = lookup_phone(phone);
        if (mapped) {
            /* Per CODE POINT: the voice tokenises ɑ, ː and r separately, so
             * "ɑːr" must arrive as three symbols, not one — and certainly not
             * as five raw bytes. */
            for (const char *p = mapped; *p; ) {
                size_t len = utf8_seq_len(p);
                char *sym = dup_range(p, len);
                if (!sym) break;
                conv.ipa[conv.ipa_count++] = sym;
                p += len;
            }
            continue;
        }

        int seen = 0;
        for (size_t u = 0; u < conv.unmapped_count; u++)
            if (strcmp(conv.unmapped[u], phone) == 0) { seen = 1; break; }
        if (!seen) {
            char *copy = dup_range(phone, strlen(phone));
            if (copy) conv.unmapped[conv.unmapped_count++] = copy;
        }
    }

    return conv;
}

void circle_voice_conversion_free(circle_voice_conversion *conv)
{
    if (!conv) return;
    if (conv->ipa) {
        for (size_t i = 0; i < conv->ipa_count; i++) free(conv->ipa[i]);
        free(conv->ipa);
    }
    if (conv->unmapped) {
        for (size_t i = 0; i < conv->unmapped_count; i++) free(conv->unmapped[i]);
        free(conv->unmapped);
    }
    memset(conv, 0, sizeof(*conv));
}

int circle_voice_xsampa_can_say_all(const char *const *xsampa, size_t count)
{
    for (size_t i = 0; i < count; i++) {
        const char *p = xsampa[i];
        if (!p || is_blank(p)) continue;
        if (!lookup_phone(p)) return 0;
    }
    return 1;
}

/* ------------------------------------------------------------------------ */
/* SentencePiece unigram                                                     */
/* ------------------------------------------------------------------------ */

/*
 * Cost charged for falling back to raw bytes.
 *
 * Any finite penalty works, because fallback only ever competes with "no path
 * at all". It must be worse than a real piece so the lattice never prefers it
 * where a piece exists, and finite so a path always exists.
 */
#define FALLBACK_PENALTY 10.0f
#define UNREACHABLE (-1e18f)

struct circle_voice_sp {
    const circle_voice_sp_entry *entries;
    size_t count;
    size_t max_piece_codepoints;
};

circle_voice_sp *circle_voice_sp_new(const circle_voice_sp_entry *entries, size_t count)
{
    circle_voice_sp *sp = (circle_voice_sp *)calloc(1, sizeof(circle_voice_sp));
    if (!sp) return NULL;
    sp->entries = entries;
    sp->count = count;
    sp->max_piece_codepoints = 1;
    for (size_t i = 0; i < count; i++) {
        size_t n = utf8_count(entries[i].piece);
        if (n > sp->max_piece_codepoints) sp->max_piece_codepoints = n;
    }
    return sp;
}

void circle_voice_sp_free(circle_voice_sp *sp) { free(sp); }

static const circle_voice_sp_entry *sp_find(const circle_voice_sp *sp,
                                            const char *piece, size_t len)
{
    for (size_t i = 0; i < sp->count; i++) {
        const char *p = sp->entries[i].piece;
        if (strlen(p) == len && memcmp(p, piece, len) == 0) return &sp->entries[i];
    }
    return NULL;
}

size_t circle_voice_sp_encode(const circle_voice_sp *sp, const char *text,
                              int *out_ids, size_t out_capacity)
{
    if (!sp || !text || !*text) return 0;

    /* SentencePiece's own normalisation: spaces become U+2581, with one
     * prepended so the first word is marked word-initial too.
     *
     * NFKC IS NOT APPLIED. C has no Unicode normaliser in the standard library
     * and this port will not vendor one for a step no fixture exercises. On
     * already-normalised input this is byte-identical to the reference; on
     * denormalised input it can differ. Recorded here rather than hidden.
     */
    static const char SEP[] = "▁";
    const size_t sep_len = sizeof(SEP) - 1;

    size_t text_len = strlen(text);
    size_t buf_cap = text_len * sep_len + sep_len + 1;
    char *norm = (char *)malloc(buf_cap);
    if (!norm) return 0;

    size_t w = 0;
    memcpy(norm + w, SEP, sep_len); w += sep_len;
    for (size_t i = 0; i < text_len; i++) {
        if (text[i] == ' ') { memcpy(norm + w, SEP, sep_len); w += sep_len; }
        else norm[w++] = text[i];
    }
    norm[w] = '\0';

    /* Byte offset of each code point boundary, so the lattice indexes code
     * points while the pieces stay plain byte ranges. */
    size_t n = utf8_count(norm);
    size_t *off = (size_t *)malloc((n + 1) * sizeof(size_t));
    if (!off) { free(norm); return 0; }
    {
        size_t k = 0, b = 0;
        while (norm[b]) { off[k++] = b; b += utf8_seq_len(norm + b); }
        off[n] = w;
    }

    float *best = (float *)malloc((n + 1) * sizeof(float));
    size_t *from = (size_t *)calloc(n + 1, sizeof(size_t));
    int *piece_id = (int *)malloc((n + 1) * sizeof(int));
    int *has_piece = (int *)calloc(n + 1, sizeof(int));
    if (!best || !from || !piece_id || !has_piece) {
        free(norm); free(off); free(best); free(from); free(piece_id); free(has_piece);
        return 0;
    }
    for (size_t i = 0; i <= n; i++) { best[i] = UNREACHABLE; piece_id[i] = -1; }
    best[0] = 0.0f;

    for (size_t i = 0; i < n; i++) {
        if (best[i] <= UNREACHABLE / 2.0f) continue;

        size_t limit = sp->max_piece_codepoints;
        if (n - i < limit) limit = n - i;
        for (size_t len = 1; len <= limit; len++) {
            const char *start = norm + off[i];
            size_t blen = off[i + len] - off[i];
            const circle_voice_sp_entry *e = sp_find(sp, start, blen);
            if (!e) continue;
            float score = best[i] + e->score;
            if (score > best[i + len]) {
                best[i + len] = score;
                from[i + len] = i;
                piece_id[i + len] = e->id;
                has_piece[i + len] = 1;
            }
        }

        /* Byte fallback for this ONE code point, so no input is ever silent. */
        size_t end = i + 1;
        float fallback = best[i] - FALLBACK_PENALTY;
        if (fallback > best[end]) {
            best[end] = fallback;
            from[end] = i;
            has_piece[end] = 0;
        }
    }

    /* Backtrack into a scratch buffer, then reverse. */
    int *rev = (int *)malloc((n * 4 + 1) * sizeof(int));
    size_t rev_count = 0;
    if (rev) {
        size_t i = n;
        while (i > 0) {
            size_t start = from[i];
            if (has_piece[i]) {
                rev[rev_count++] = piece_id[i];
            } else {
                /* BACKWARDS, because this whole list is built backwards. The
                 * lattice is walked from the end and reversed at the bottom, so
                 * bytes pushed in forward order come out byte-reversed: é is
                 * UTF-8 C3 A9 and would be emitted A9 C3. Nothing crashes —
                 * those are real pieces with real ids — so the model simply
                 * says a different character. */
                size_t b_end = off[i], b_start = off[start];
                for (size_t b = b_end; b > b_start; b--) {
                    unsigned char byte = (unsigned char)norm[b - 1];
                    char key[9];
                    snprintf(key, sizeof(key), "<0x%02X>", byte);
                    const circle_voice_sp_entry *e = sp_find(sp, key, strlen(key));
                    if (e) rev[rev_count++] = e->id;
                }
            }
            i = start;
        }

        for (size_t k = 0; k < rev_count && k < out_capacity; k++)
            out_ids[k] = rev[rev_count - 1 - k];
    }

    size_t result = rev_count;
    free(rev); free(norm); free(off); free(best); free(from); free(piece_id); free(has_piece);
    return result;
}

File: rust/src/voice_xsampa.rs
//! voice_xsampa — X-SAMPA → IPA, and SentencePiece unigram encoding.
//!
//! Port of `src/CircleAI.Voice/XsampaToIpa.cs` and
//! `src/CircleAI.Voice/SentencePieceUnigram.cs`.
//!
//! Parity is asserted against `fixtures/voice_xsampa_to_ipa.json` and
//! `fixtures/voice_sentencepiece_unigram.json`, which the C# reference
//! generates. If this file and those files disagree, one of them is wrong and
//! the test names the case.

use std::collections::HashMap;
use std::sync::OnceLock;

/// Every phone in the NCHLT Afrikaans dictionary, mapped to IPA.
///
/// Derived from the corpus, not from memory: exactly the distinct phones in
/// `nchlt_afr.dict`, with every IPA character checked against the target
/// voice's own token table before the table was written.
fn table() -> &'static HashMap<&'static str, &'static str> {
    static TABLE: OnceLock<HashMap<&'static str, &'static str>> = OnceLock::new();
    TABLE.get_or_init(|| {
        HashMap::from([
            // Vowels
            ("a", "a"), ("A:", "ɑː"), ("A:r", "ɑːr"),
            ("E", "ɛ"), ("O", "ɔ"), ("@", "ə"),
            ("i", "i"), ("u", "u"), ("y", "y"),
            ("9", "œ"), ("2:", "øː"), ("{", "æ"),

            // Diphthongs — NCHLT gives one token, the voice wants both elements.
            ("9y", "œy"), ("@i", "əi"), ("@u", "əu"),
            ("i@", "iə"), ("u@", "uə"),

            // Consonants
            ("b", "b"), ("d", "d"), ("f", "f"),
            // U+0261 LATIN SMALL LETTER SCRIPT G — the IPA letter, NOT ASCII
            // 'g'. The voice's vocabulary carries ɡ; a plain 'g' would miss and
            // be dropped.
            ("g", "\u{0261}"),
            ("j", "j"), ("k", "k"), ("l", "l"),
            ("m", "m"), ("n", "n"), ("N", "ŋ"),
            ("p", "p"), ("r", "r"), ("s", "s"),
            ("S", "ʃ"), ("t", "t"), ("v", "v"),
            ("w", "w"), ("x", "x"), ("z", "z"),
            ("Z", "ʒ"),

            // APPROXIMATION, DELIBERATE AND THE ONLY ONE. X-SAMPA h\ is ɦ, the
            // voiced glottal fricative Afrikaans uses in "hond". This voice's
            // vocabulary has no ɦ, only h. Voicing is lost; place and manner are
            // right, so the word stays recognisable.
            ("h\\", "h"),
        ])
    })
}

/// The IPA symbols for `xsampa`, plus the phones that could not be mapped.
///
/// Returned together rather than stashed in a static, because an unmapped phone
/// produces NO SOUND and the audio is merely shorter — every acoustic measure
/// still passes. A caller that cannot see the misses cannot refuse.
///
/// LONGEST MATCH ON WHOLE TOKENS. Several entries are multi-character (`A:r`,
/// `@i`, `9y`) and NCHLT emits them as single tokens; matching on the token —
/// never character by character — is what keeps `A:r` from becoming `A` + `:` + `r`.
pub fn xsampa_to_ipa(xsampa: &[&str]) -> (Vec<String>, Vec<String>) {
    let map = table();
    let mut ipa = Vec::with_capacity(xsampa.len() + 8);
    let mut unmapped: Vec<String> = Vec::new();

    for phone in xsampa {
        if phone.trim().is_empty() {
            continue;
        }
        if let Some(mapped) = map.get(phone) {
            // Per-char: the voice tokenises ɑ, ː and r separately, so "ɑːr" must
            // arrive as three symbols, not one.
            for ch in mapped.chars() {
                ipa.push(ch.to_string());
            }
            continue;
        }
        if !unmapped.iter().any(|u| u == phone) {
            unmapped.push((*phone).to_string());
        }
    }

    (ipa, unmapped)
}

/// True when every phone in `xsampa` has a mapping.
pub fn xsampa_can_say_all(xsampa: &[&str]) -> bool {
    let map = table();
    xsampa
        .iter()
        .filter(|p| !p.trim().is_empty())
        .all(|p| map.contains_key(p))
}

/// The X-SAMPA phones this table knows — for tests and diagnostics.
pub fn xsampa_known_phones() -> Vec<String> {
    table().keys().map(|k| (*k).to_string()).collect()
}

// ─────────────────────────────────────────────────────────────────────────────
// SentencePiece unigram
// ─────────────────────────────────────────────────────────────────────────────

/// Cost charged for falling back to raw bytes.
///
/// Any finite penalty works, because fallback only ever competes with "no path
/// at all". It must be worse than a real piece so the lattice never prefers it
/// where a piece exists, and finite so a path always exists.
const FALLBACK_PENALTY: f32 = 10.0;

/// SentencePiece unigram tokeniser — Viterbi over the piece lattice, with byte
/// fallback.
pub struct SentencePieceUnigram {
    ids: HashMap<String, i64>,
    scores: HashMap<String, f32>,
    max_piece_length: usize,
}

impl SentencePieceUnigram {
    /// Build from piece→id and piece→score maps.
    pub fn new(ids: HashMap<String, i64>, scores: HashMap<String, f32>) -> Self {
        let max_piece_length = ids.keys().map(|k| k.chars().count()).max().unwrap_or(1);
        Self { ids, scores, max_piece_length }
    }

    /// Number of pieces in the vocabulary.
    pub fn count(&self) -> usize {
        self.ids.len()
    }

    /// Encode text to token ids.
    ///
    /// VITERBI, NOT GREEDY LONGEST-MATCH. Unigram scores are not monotone in
    /// piece length — a long piece can score worse than the two short pieces
    /// covering the same span — so greedy silently produces plausible-but-wrong
    /// segmentations.
    pub fn encode(&self, text: &str) -> Vec<i64> {
        if text.is_empty() {
            return Vec::new();
        }

        // SentencePiece's own normalisation: spaces become U+2581, with one
        // prepended so the first word is marked word-initial too.
        //
        // NFKC is NOT applied here. The C# reference calls Normalize(FormKC) and
        // Rust has no stdlib normaliser; rather than pull in unicode-normalization
        // for a step no fixture exercises, this port stays byte-identical on
        // already-normalised input and is honest about the gap. Feeding
        // denormalised text to this port and to C# can differ — see the module
        // note in docs/CONTRACTS.md.
        let mut normalised = String::with_capacity(text.len() + 4);
        normalised.push('▁');
        for ch in text.chars() {
            normalised.push(if ch == ' ' { '▁' } else { ch });
        }

        // CHARS, NOT BYTES. Indexing by byte would let a piece boundary land
        // mid-codepoint, producing pieces that match nothing and byte-fallback
        // output that decodes to a different character.
        let chars: Vec<char> = normalised.chars().collect();
        let n = chars.len();

        const UNREACHABLE: f32 = -1e18;
        let mut best = vec![UNREACHABLE; n + 1];
        let mut from_index = vec![0usize; n + 1];
        let mut piece: Vec<Option<String>> = vec![None; n + 1];
        best[0] = 0.0;

        for i in 0..n {
            if best[i] <= UNREACHABLE / 2.0 {
                continue;
            }

            let limit = self.max_piece_length.min(n - i);
            for len in 1..=limit {
                let candidate: String = chars[i..i + len].iter().collect();
                if !self.ids.contains_key(&candidate) {
                    continue;
                }
                let score = best[i] + self.scores.get(&candidate).copied().unwrap_or(0.0);
                if score > best[i + len] {
                    best[i + len] = score;
                    from_index[i + len] = i;
                    piece[i + len] = Some(candidate);
                }
            }

            // Byte fallback for this ONE char, so no input is ever silent.
            let end = i + 1;
            let fallback = best[i] - FALLBACK_PENALTY;
            if fallback > best[end] {
                best[end] = fallback;
                from_index[end] = i;
                piece[end] = None;
            }
        }

        let mut reversed: Vec<i64> = Vec::with_capacity(n);
        let mut i = n;
        while i > 0 {
            let start = from_index[i];
            match &piece[i] {
                Some(p) => {
                    if let Some(id) = self.ids.get(p) {
                        reversed.push(*id);
                    }
                }
                None => {
                    // BACKWARDS, because this whole list is built backwards. The
                    // lattice is walked from the end and flipped once at the
                    // bottom, so a multi-byte character pushed in forward order
                    // comes out byte-reversed: é is UTF-8 C3 A9 and would be
                    // emitted A9 C3. Nothing panics — those are real pieces with
                    // real ids — so the model simply says a different character.
                    let raw: String = chars[start..i].iter().collect();
                    for b in raw.as_bytes().iter().rev() {
                        if let Some(id) = self.ids.get(&format!("<0x{b:02X}>")) {
                            reversed.push(*id);
                        }
                    }
                }
            }
            i = start;
        }

        reversed.reverse();
        reversed
    }
}
